File: DAIRemoteApp/app/src/main/java/com/example/dairemote_app/ConnectionMonitor.java
package com.example.dairemote_app;

import android.os.Handler;
import android.os.Looper;

import java.net.DatagramSocket;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ConnectionMonitor {
    public static ConnectionMonitor connectionMonitorInstance;
    private ConnectionManager connectionManager;
    private static ExecutorService heartbeatExecutorService;
    private Handler handler;
    private Runnable heartbeatService;
    private static String heartbeatResponse;
    private boolean serviceRunning = false;
    private SocketManager heartbeatSocket;

    public ConnectionMonitor(ConnectionManager manager) {
        SetHeartbeatSocket(new SocketManager(ConnectionManager.GetInetAddress(), ConnectionManager.GetPort()));
        SetConnectionManager(manager);
        handler = new Handler(Looper.getMainLooper());
        heartbeatService = new Runnable() {
            @Override
            public void run() {
                // Send heartbeat
                if (SendHeartbeat()) {
                    handler.postDelayed(this, 2000); // Repeat every 2 seconds
                } else {
                    SetServiceRunning(false);
                    GetConnectionManager().SetConnectionEstablished(false);
                }
            }
        };

        this.heartbeatExecutorService = Executors.newCachedThreadPool();
    }

    public static ConnectionMonitor GetInstance(ConnectionManager manager) {
        if(connectionMonitorInstance == null) {
            connectionMonitorInstance = new ConnectionMonitor(manager);
        }
        return connectionMonitorInstance;
    }

    public void SetConnectionManager(ConnectionManager manager) {
        this.connectionManager = manager;
    }

    public ConnectionManager GetConnectionManager() {
        return this.connectionManager;
    }

    public void SetHeartbeatSocket(SocketManager socket) { this.heartbeatSocket = socket; }

    public SocketManager GetHeartbeatSocket() { return this.heartbeatSocket; }

    public static void SetHeartbeatResponse(String response) {
        heartbeatResponse = response;
    }

    public static String GetHeartbeatResponse() {
        return heartbeatResponse;
    }

    public boolean IsHeartbeatRunning() {
        return GetServiceRunning();
    }

    public void SetServiceRunning(boolean serviceRunning) { this.serviceRunning = serviceRunning; };

    public boolean GetServiceRunning() { return this.serviceRunning; };

    public void StartHeartbeat() {
        if(!GetServiceRunning()) {
            handler.post(heartbeatService);
            SetServiceRunning(true);
        }
    }

    public void StartHeartbeat(int delay) {
        if(!GetServiceRunning()) {
            handler.postDelayed(heartbeatService, delay);
            SetServiceRunning(true);
        }
    }

    public boolean SendHeartbeat() {
        if (heartbeatExecutorService.isShutdown()) {
            return false;
        }

        // Retry heartbeat 5 times
        for (int attempt = 0; attempt < 2; attempt++) {
            CompletableFuture<Boolean> future = CompletableFuture.supplyAsync(() -> {
                try {
                    GetHeartbeatSocket().SendData("DroidHeartBeat");
                    SetHeartbeatResponse(GetHeartbeatSocket().WaitForResponse(3000));

                    if (GetHeartbeatResponse().equalsIgnoreCase("HeartBeat Ack")) {
                        return true;
                    }
                } catch (Exception ignored) {
                }
                return false;
            }, heartbeatExecutorService);

            try {
                boolean result = future.get();
                if (result) {
                    return true;
                }
            } catch (Exception e) {
                e.printStackTrace();
            }

            try {
                Thread.sleep(125);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        return false;
    }

    public void StopHeartbeat() {
        handler.removeCallbacks(heartbeatService);
        SetServiceRunning(false);
    }

    public void ShutDownHeartbeat() {
        StopHeartbeat();
        if (heartbeatExecutorService != null && !heartbeatExecutorService.isShutdown()) {
            heartbeatExecutorService.shutdownNow();
        }
        GetHeartbeatSocket().CloseSocket();
    }
}

File: DAIRemoteApp/app/src/main/java/com/example/dairemote_app/utils/ConnectionMonitor.kt
package com.example.dairemote_app.utils

import android.os.Handler
import android.os.Looper
import com.example.dairemote_app.viewmodels.ConnectionViewModel
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

private const val HEARTBEAT_INTERVAL_MS = 2000L
private const val HEARTBEAT_TIMEOUT_MS: Int = 250
private const val MAX_MISSED_HEARTBEATS = 3

class ConnectionMonitor(manager: ConnectionManager, private val viewModel: ConnectionViewModel) {
    private lateinit var connectionManager: ConnectionManager
    private var connectionView: ConnectionViewModel
    private lateinit var handler: Handler
    private var serviceRunning = false
    private lateinit var heartbeatSocket: SocketManager
    private var missedHeartbeats = 0

    init {
        setConnectionManager(manager)
        connectionView = viewModel
    }

    private fun setConnectionManager(manager: ConnectionManager) {
        this.connectionManager = manager
    }

    fun getConnectionManager(): ConnectionManager {
        return this.connectionManager
    }

    private fun setHeartbeatSocket(socket: SocketManager) {
        this.heartbeatSocket = socket
    }

    private fun getHeartbeatSocket(): SocketManager {
        return this.heartbeatSocket
    }

    private val heartbeatService = object : Runnable {
        override fun run() {
            if (!serviceRunning) return

            try {
                if (sendHeartbeat()) {
                    missedHeartbeats = 0
                    handler.postDelayed(this, HEARTBEAT_INTERVAL_MS)
                } else {
                    missedHeartbeats++
                    if (missedHeartbeats >= MAX_MISSED_HEARTBEATS) {
                        handleConnectionLoss()
                    } else {
                        handler.postDelayed(this, 250)
                    }
                }
            } catch (e: Exception) {
                handleConnectionLoss()
            }
        }
    }

    private fun handleConnectionLoss() {
        viewModel.updateConnectionState(false)
        shutDownHeartbeat()
        getConnectionManager().resetConnectionManager()
        getConnectionManager().closeUDPSocket()
    }

    private fun initHeartbeat() {
        handler = Handler(Looper.getMainLooper())
        heartbeatExecutorService = Executors.newCachedThreadPool()
    }

    private fun setServiceRunning(serviceRunning: Boolean) {
        this.serviceRunning = serviceRunning
    }

    private fun getServiceRunning(): Boolean {
        return this.serviceRunning
    }

    fun startHeartbeat(delay: Int) {
        if (!getServiceRunning()) {
            setHeartbeatSocket(
                SocketManager(
                    ConnectionManager.getInetAddress(),
                    ConnectionManager.getPort()
                )
            )
            initHeartbeat()
            handler.postDelayed(heartbeatService, delay.toLong())
            setServiceRunning(true)
        }
    }

    fun sendHeartbeat(): Boolean {
        if (heartbeatExecutorService.isShutdown) {
            return false
        }

        val future = CompletableFuture.supplyAsync(
            {
                try {
                    getHeartbeatSocket().sendData("DroidHeartBeat")
                    setHeartbeatResponse(
                        getHeartbeatSocket().waitForResponse(
                            HEARTBEAT_TIMEOUT_MS
                        )
                    )

                    if (getHeartbeatResponse().equals("HeartBeat Ack", ignoreCase = true)) {
                        return@supplyAsync true
                    }
                } catch (ignored: Exception) {
                }
                false
            }, heartbeatExecutorService
        )

        try {
            val result = future.get()
            if (result) {
                return true
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        try {
            Thread.sleep(125)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

        return false
    }

    private fun stopHeartbeat() {
        handler.removeCallbacks(heartbeatService)
        setServiceRunning(false)
    }

    fun shutDownHeartbeat() {
        if (getServiceRunning()) {
            stopHeartbeat()
            if (!heartbeatExecutorService.isShutdown) {
                heartbeatExecutorService.shutdownNow()
            }
            getHeartbeatSocket().closeSocket()
        }
    }

    companion object {
        private var connectionMonitorInstance: ConnectionMonitor? = null
        private lateinit var heartbeatExecutorService: ExecutorService
        private var heartbeatResponse: String? = null

        fun getInstance(manager: ConnectionManager, viewModel: ConnectionViewModel): ConnectionMonitor? {
            if (connectionMonitorInstance == null) {
                connectionMonitorInstance = ConnectionMonitor(manager, viewModel)
            }
            return connectionMonitorInstance
        }

        fun setHeartbeatResponse(response: String?) {
            heartbeatResponse = response
        }

        fun getHeartbeatResponse(): String? {
            return heartbeatResponse
        }
    }
}
